Signal/src/ViewControllers/SelectThreadViewController.m

```diff
@@ -0,0 +1,428 @@
+        } else if ([thread isKindOfClass:[TSContactThread class]]) {
+            TSContactThread *contactThread = (TSContactThread *)thread;
+            if (formattedNumber.length > 0 && [contactThread.contactIdentifier containsString:formattedNumber]) {
+                [result addObject:thread];
+            }
+        }
+    }
+    return result;
+}
+
+// TODO: Move this to contacts view helper.
+- (NSArray<Contact *> *)filteredContactsWithSearchText
+{
+    // We don't want to show a 1:1 thread with Alice and Alice's contact,
+    // so we de-duplicate by recipientId.
+    NSArray<TSThread *> *threads = self.threadViewHelper.threads;
+    NSMutableSet *contactIdsToIgnore = [NSMutableSet new];
+    for (TSThread *thread in threads) {
+        if ([thread isKindOfClass:[TSContactThread class]]) {
+            TSContactThread *contactThread = (TSContactThread *)thread;
+            [contactIdsToIgnore addObject:contactThread.contactIdentifier];
+        }
+    }
+
+    NSString *searchString = [self.searchBar text];
+
+    NSArray *nonRedundantContacts =
+        [self.contacts filteredArrayUsingPredicate:[NSPredicate predicateWithBlock:^BOOL(Contact *contact,
+                                                       NSDictionary<NSString *, id> *_Nullable bindings) {
+            return ![contactIdsToIgnore containsObject:contact.textSecureIdentifiers.firstObject];
+        }]];
+
+    // TODO: Move this to contacts view helper.
+    OWSContactsSearcher *contactsSearcher = [[OWSContactsSearcher alloc] initWithContacts:nonRedundantContacts];
+    NSArray<Contact *> *filteredContacts = [contactsSearcher filterWithString:searchString];
+
+    return filteredContacts;
+}
+
+#pragma mark - Contacts and Blocking
+
+- (void)blockedPhoneNumbersDidChange:(id)notification
+{
+    dispatch_async(dispatch_get_main_queue(), ^{
+        _blockedPhoneNumberSet = [NSSet setWithArray:[_blockingManager blockedPhoneNumbers]];
+
+        [self updateContacts];
+    });
+}
+
+- (void)signalRecipientsDidChange:(NSNotification *)notification
+{
+    dispatch_async(dispatch_get_main_queue(), ^{
+        [self updateContacts];
+    });
+}
+
+- (void)updateContacts
+{
+    OWSAssert([NSThread isMainThread]);
+
+    self.contacts = [self filteredContacts];
+    [self updateTableContents];
+}
+
+- (BOOL)isContactBlocked:(Contact *)contact
+{
+    if (contact.parsedPhoneNumbers.count < 1) {
+        // Hide contacts without any valid phone numbers.
+        return NO;
+    }
+    
+    for (PhoneNumber *phoneNumber in contact.parsedPhoneNumbers) {
+        if ([_blockedPhoneNumberSet containsObject:phoneNumber.toE164]) {
+            return YES;
+        }
+    }
+    
+    return NO;
+}
+
+- (BOOL)isRecipientIdBlocked:(NSString *)recipientId
+{
+    OWSAssert(recipientId.length > 0);
+    
+    return [_blockedPhoneNumberSet containsObject:recipientId];
+}
+
+- (BOOL)isContactHidden:(Contact *)contact
+{
+    if (contact.parsedPhoneNumbers.count < 1) {
```

Comment:
nit: this is a little confusing, since `self.contacts` is populated with `contactsManager.signalContacts`, which necessarily have a signal id. Having checks like this muddy the waters as to whose responsibility it is to make sure signalContacts are valid signalContacts.